google-cloud-firestore/src/main/java/com/google/cloud/firestore/AggregateQuery.java

```diff
@@ -112,52 +143,156 @@ private final class AggregateQueryResponseDeliverer {
       this.startTimeNanos = startTimeNanos;
     }
 
-    ApiFuture<AggregateQuerySnapshot> getFuture() {
+    @Nullable
+    ByteString getTransactionId() {
+      return transactionId;
+    }
+
+    @Nullable
+    com.google.protobuf.Timestamp getReadTime() {
+      return readTime;
+    }
+
+    long getStartTimeNanos() {
+      return startTimeNanos;
+    }
+
+    @Nullable
+    ExplainOptions getExplainOptions() {
+      return null;
+    }
+
+    boolean futureNotCompleted() {
+      return futureCompleted.compareAndSet(false, true);
+    }
+
+    ApiFuture<T> getFuture() {
       return future;
     }
 
-    void deliverResult(@Nonnull Map<String, Value> data, Timestamp readTime) {
-      if (isFutureCompleted.compareAndSet(false, true)) {
-        Map<String, Value> mappedData = new HashMap<>();
-        data.forEach((serverAlias, value) -> mappedData.put(aliasMap.get(serverAlias), value));
-        future.set(new AggregateQuerySnapshot(AggregateQuery.this, readTime, mappedData));
-      }
+    void setFuture(T value) {
+      future.set(value);
     }
 
     void deliverError(Throwable throwable) {
-      if (isFutureCompleted.compareAndSet(false, true)) {
+      if (futureNotCompleted()) {
```

Comment:
The `setException` has built in control to make sure it is only called once. From docs:

> When a call to this method returns, the Future is guaranteed to be done only if the call was accepted (in which case it returns true).

The `AtomicBoolean` is not required.

Reply:
Done